### PR TITLE
cleanup / refactoring

### PR DESCRIPTION
**Pull request by alcohol**

### CODE REVIEW

File: config.go [hunk 1]
```diff
@@ -16,20 +16,18 @@ type JsonConfig struct {
...
+		return err
 	}
 
 	defer h.Close()
 
-	c := JsonConfig{}
-
-	if err = json.NewDecoder(h).Decode(&c); err != nil {
-		return JsonConfig{}, err
+	if err = json.NewDecoder(h).Decode(&this); err != nil {
```
Comment (SpComb) [inline config.go]: The `&this` is probably a bad idea, too many pointers.

You want json to replace the empty `*this` map with a new map.

File: config.json.dist [hunk 2]
```diff
@@ -6,5 +6,5 @@
   "username": "composer",
   "password": "",
   "cmd-prefix": "!",
-  "factoids": "factoids.json"
+  "factoids": "/etc/irc-bot/factoids.json"
```
Comment (SpComb) [inline config.json.dist]: Writing variable state to files under `/etc` is always a bad idea. That's what `/var` is for.

File: bot.go [hunk 3]
```diff
@@ -13,15 +14,42 @@ import (
...
+func (this *Bot) Config(filename string) error {
+	h, err := os.Open(filename)
+
+	if err != nil {
+		return err
+	}
+
+	defer h.Close()
+
+	if err = json.NewDecoder(h).Decode(&this); err != nil {
```
Comment (SpComb) [inline bot.go]: Too much pointer indirection with `&this`.

File: factoids.go [hunk 4]
```diff
@@ -7,57 +7,26 @@ import (
...
+		return err
 	}
 
 	defer h.Close()
 
-	f := factoids{filename: filename}
-
-	if err = json.NewDecoder(h).Decode(&f.contents); err != nil {
-		return factoids{filename, make(map[string]string, 0)}, nil
+	if err = json.NewDecoder(h).Decode(&this); err != nil {
```
Comment (SpComb) [inline factoids.go]: The `&this` is probably a bad idea, too many pointers, should just be `.Decode(this)`.

You effectively want json to replace the empty-valued `*this` map with a new `*this = make(map[...]...)`

**Comment by SpComb** [comment]
http://softwareengineering.stackexchange.com/questions/286406/use-of-this-in-golang
http://stackoverflow.com/questions/23482068/in-go-is-naming-the-receiver-variable-self-misleading-or-good-practise

File: bot.go [hunk 5]
```diff
@@ -14,18 +15,41 @@ import (
 )
 
 type Bot struct {
-	Config JsonConfig
+	Server    string   `json:"server"`
```
Comment (mrosset) [inline bot.go]: This could be a Server struct

File: bot.go [hunk 6]
```diff
@@ -14,18 +15,41 @@ import (
...
+	Port      int      `json:"port"`
+	Channels  []string `json:"channels"`
```
Comment (mrosset) [inline bot.go]: Channels and ports, can be fields on the Server struct

File: bot.go [hunk 7]
```diff
@@ -13,33 +14,69 @@ import (
...
+func (bot *Bot) Configure(filename string) error {
+	handle, err := os.Open(filename)
+
+	if err != nil {
+		return err
+	}
+
+	defer handle.Close()
+
+	if err = json.NewDecoder(handle).Decode(bot); err != nil {
```
Comment (mrosset) [inline bot.go]: you can return from here.

return json.NewDecoder(handle).Decode(bot)

**Pull request merged**